[ZEPPELIN-3377] Passing Z variables to JDBC interpreter

sanjaydasgupta opened the pull request:
### What is this PR for?
This PR enables the interpolation of ZeppelinContext objects into the paragraph text of JDBC cells. It also introduces a new interpreter-level configuration parameter named _zeppelin.jdbc.interpolation_. This new parameter is _false_ by default, and must be set to _true_ to enable object interpolation. The default value of _false_ guarantees backward compatibility for users who are not aware of the new feature.

The implementation takes the same approach that was followed in [PR-2898](https://github.com/apache/zeppelin/pull/2898).

I have also taken the liberty to correct a preexisting error in the description of the use of Dynamic Forms in the associated documentation (_jdbc.md_).

### What type of PR is it?
[Feature]

### Todos
* [ ] - Task

### What is the Jira issue?
https://issues.apache.org/jira/browse/ZEPPELIN-3377

### How should this be tested?
CI Pass. 

The code in this PR merely causes the JDBC interpreter to "opt-in" to the implementation already existing in the `Interpreter` base class - described in [PR-2898](https://github.com/apache/zeppelin/pull/2898). The unit tests necessary are already present in PR-2898

### Screenshots (if appropriate)

### Questions:
* Does the licenses files need update? No
* Is there breaking changes for older versions? No
* Does this needs documentation? Yes, documentation has been added to the file _jdbc.md_. I have also taken the liberty to correct a preexisting error in the description of the use of Dynamic Forms in the associated documentation.

felixcheung commented on docs/interpreter/jdbc.md:
```diff
@@ -240,7 +244,7 @@ You can leverage [Zeppelin Dynamic Form](../usage/dynamic_form/intro.html) insid
 SELECT name, country, performer
 FROM demo.performers
-WHERE name='{{"{{performer=Sheryl Crow|Doof|Fanfarlo|Los Paranoia"}}}}'
+WHERE name='${performer=Sheryl Crow|Doof|Fanfarlo|Los Paranoia}'
```
I think this is different, right, why this change?

sanjaydasgupta replied:
Yes, this is a different thing. I have mentioned under **What is this PR for?**:

> I have also taken the liberty to correct a preexisting error in the description of the use of
> Dynamic Forms in the associated documentation (jdbc.md).

Since I was editing _jdbc.md_, I thought correcting the error would be useful for readers. The changed text works as intended under the current implementation, and is also according to the documentation for Dynamic Forms.

Please let me know if corrections should be handled in a different way. 

Thanks for noticing this update.

felixcheung replied:
ok thanks

felixcheung commented on docs/interpreter/jdbc.md:
```diff
@@ -723,6 +727,26 @@ Before Adding one of the below dependencies, check the Phoenix version first.
 </table>
 
 [Maven Repository: org.apache.tajo:tajo-jdbc](https://mvnrepository.com/artifact/org.apache.tajo/tajo-jdbc)
+## Object Interpolation
```
I think you need an empty line before `##` or it will not be treated as header properly

sanjaydasgupta replied:
Added preceding blank line.

felixcheung commented on docs/interpreter/jdbc.md:
````diff
@@ -723,6 +727,26 @@ Before Adding one of the below dependencies, check the Phoenix version first.
+```
+
+####In later JDBC cell:
+```
````
add `sql` - see [L230](https://github.com/apache/zeppelin/pull/2903/files#diff-12ee03bf05e397a25720e23c90890661R230)

sanjaydasgupta replied:
Added the 'sql'

felixcheung commented on docs/interpreter/jdbc.md:
````diff
@@ -723,6 +727,26 @@ Before Adding one of the below dependencies, check the Phoenix version first.
+
+####In later JDBC cell:
+```
+%jdbc_interpreter_name select * from patents_list where 
````
newline after `%jdbc_interpreter_name` for clarity & consistency with other example?

sanjaydasgupta replied:
Done.

felixcheung commented on docs/interpreter/jdbc.md:
````diff
@@ -723,6 +727,26 @@ Before Adding one of the below dependencies, check the Phoenix version first.
+####In later JDBC cell:
+```
+%jdbc_interpreter_name select * from patents_list where 
+    priority_country = '{country_code}' and filing_date like '2015-__-__'
````
is `_` in `2015-__-__` intentional?
perhaps better if the example is `2015-%`? or just pick `2015-01-01`

sanjaydasgupta replied:
I've preferred '2015-%' as using '2015-01-01' would retrieve a too small sample - and not be a realistic example.

sanjaydasgupta commented:
@felixcheung I've fixed the 4 issues in the documentation file jdbc.md, thanks for the very careful review.

I have also added a unit test [JDBCInterpreterInterpolationTest.java](https://github.com/apache/zeppelin/pull/2903/files#diff-f8ebbd56b41630cac1cec76efe751d9d) which tests the following:

1) enabling and disabling the feature by setting "zeppelin.jdbc.interpolation", including the default value of false.
2) normal substitution of {...} patterns
3) escaping the pattern functionality by using {{...}}

Please also note that all the unit tests defined for this feature in the `Interpreter` base class do a thorough check of all possible uses and misuses of { and }. The above additional unit test is really more like an integration test to verify that the JDBC interpreter is inheriting and leveraging the base class's functionality correctly.

Thanks for your very helpful review.

sanjaydasgupta commented:
Apologies for missing the check style errors @felixcheung 

The style issues in _JDBCInterpreterInterpolationTest.java_ have been removed, and all the tests now pass. The other failures in the build are unrelated to this feature.

Thanks  for your review.

felixcheung commented:
still have errors on this link https://travis-ci.org/sanjaydasgupta/zeppelin/builds/363808230?

sanjaydasgupta commented:
Oops! I must have messed up somewhere!

I've corrected the style issues now @felixcheung, and the one remaining failure appears to be a build script problem unrelated to my code. 

Thanks!

sanjaydasgupta commented:
ping @felixcheung

sanjaydasgupta commented:
Closing and reopening to trigger tests.

sanjaydasgupta commented:
Hi @felixcheung please let me know if any additional changes or adjustments are needed.

The two failing tests are unrelated to the code in in this PR

Thanks.

felixcheung commented:
looks good!
merging if no more comment

sanjaydasgupta commented:
Hi @felixcheung is it possible to release this into 0.8.0?

Thanks

felixcheung commented:
I think so, @zjffdu what do you think about this in 0.8?

felixcheung commented:
merged this to master & 0.8
